### DISCUSSION TITLE
Using JSError in local Rust tests?

### DISCUSSION THREAD

**Discussion by kristian1108** [wasm-bindgen/wasm-bindgen#3572] [Q&A]
I'm new to this wasm ecosystem so apologies if this is a silly issue/question.

I'm trying to test my wasm module inside the Rust project, the same way I would do with a normal Rust crate. I'm running into an issue where I can't pass errors outside of the wasm functions that are compatible with my local architecture. For example:

```rust
#[wasm_bindgen]
pub fn foo(bar: &str) -> Result<(), JsError> {}
```

If this method returns an error, it will panic my tests with:

```
thread 'server::auth_test::test_activate_user_v2' panicked at 'function not implemented on non-wasm32 targets', /Users/kristian/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasm-bindgen-0.2.87/src/lib.rs:996:1
```

After tracing through the code, I discovered this is originating in the `JsError::new()` call. You cannot instantiate a `JSError` outside of a wasm architecture.

Is there a way to get around this so I can pass helpful errors back in my test code?

One solution is to dynamically compile in a different error. For example:

```rust 
#[cfg(target_arch = "wasm32")]
type WasmError = JsError;

#[cfg(not(target_arch = "wasm32"))]
type WasmError = anyhow::Error;

#[wasm_bindgen]
pub fn foo(bar: &str) -> Result<(), WasmError> {}
```

But this seems like a lot of headache for what I have to think is a very common problem? Thanks for your help!

**Comment by daxpedda** [comment]
I'm not entirely sure how this could be actionable by `wasm-bindgen`. `JsError` can't be used on non-Wasm targets, it doesn't make any sense. This is a low-level bindings crate after all.

How did you end up exactly in this situation? I'm surprised you are using a `#[wasm_bindgen]` function on native in the first place. I'm happy to take a look at some actual code and give some recommendations based on that.

**Reply by kristian1108** [reply]
What's the suggested way to test `#[wasm_bindgen]` functions? I was writing tests in Rust for these methods, but maybe that's not the right answer? If you write tests in Rust, you'll be running these functions on native.

**Reply by sisou** [reply]
With `wasm_bindgen_test` and `wasm-pack test`: https://rustwasm.github.io/wasm-bindgen/wasm-bindgen-test/usage.html

**Reply by daxpedda** [reply]
Apologies for not replying back then btw, apparently I missed it in my inbox somehow.

**Comment by lex-gsynergy** [comment]
An approach I am using is to use Result<Foo, String> internally in Rust, and then to only convert the String to a JsError in the API functions in between WASM and JavaScript. This way allows more things to be tested with native builds and unit tests.
